devtools/projects/ng-devtools/src/lib/application-services/theme_service.ts

```diff
@@ -6,43 +6,76 @@
-const DARK_THEME_CLASS = 'dark-theme';
-const LIGHT_THEME_CLASS = 'light-theme';
-
-@Injectable()
+@Injectable({providedIn: 'root'})
 export class ThemeService {
-  private win = inject(WINDOW);
-  private doc = inject(DOCUMENT);
-  private renderer: Renderer2;
-  readonly currentTheme = signal<Theme>(LIGHT_THEME_CLASS);
-
-  constructor(private _rendererFactory: RendererFactory2) {
-    this.renderer = this._rendererFactory.createRenderer(null, null);
-    this.toggleDarkMode(this.prefersDarkMode);
+  private readonly win = inject(WINDOW);
+  private readonly doc = inject(DOCUMENT);
+  private readonly settings = inject(Settings);
+  private readonly rendererFactory = inject(RendererFactory2);
+  private readonly renderer = this.rendererFactory.createRenderer(null, null);
+
+  private readonly preferredTheme = signal<ThemeUi>(this.prefersDarkMode ? 'dark' : 'light');
+
+  currentTheme = computed<ThemeUi>(() => {
+    const theme = this.settings.theme();
+    if (theme === 'system') {
+      return this.preferredTheme();
+    }
+    return theme;
+  });
+
+  constructor() {
+    effect(() => {
+      const theme = this.currentTheme();
+      this.updateThemeClass(theme);
+    });
   }
 
   private get prefersDarkMode(): boolean {
     return this.win.matchMedia && this.win.matchMedia('(prefers-color-scheme: dark)').matches;
   }
 
   toggleDarkMode(isDark: boolean): void {
-    const removeClass = isDark ? LIGHT_THEME_CLASS : DARK_THEME_CLASS;
-    const addClass = !isDark ? LIGHT_THEME_CLASS : DARK_THEME_CLASS;
-    this.renderer.removeClass(this.doc.documentElement, removeClass);
-    this.renderer.addClass(this.doc.documentElement, addClass);
-    this.currentTheme.set(addClass);
+    const theme: ThemeUi = isDark ? 'dark' : 'light';
+    this.settings.theme.set(theme);
   }
 
   initializeThemeWatcher(): void {
-    this.win.matchMedia('(prefers-color-scheme: dark)').addEventListener('change', (e) => {
-      this.toggleDarkMode(this.prefersDarkMode);
+    this.win.matchMedia('(prefers-color-scheme: dark)').addEventListener('change', () => {
+      // We don't need to keep track of the preferred theme if `theme` != 'system'.
+      // On the contrary, we wouldn't like to trigger `currentTheme` recalc since it's redundant.
+      if (this.settings.theme() === 'system') {
+        const prefers: ThemeUi = this.prefersDarkMode ? 'dark' : 'light';
+        this.preferredTheme.set(prefers);
+      }
+    });
+  }
+
+  private updateThemeClass(theme: ThemeUi) {
+    this.removeThemeClass();
+
+    const themeClass = theme === 'dark' ? DARK_THEME_CLASS : LIGHT_THEME_CLASS;
+    this.renderer.addClass(this.doc.documentElement, themeClass);
+  }
+
+  private removeThemeClass() {
+    const htmlEl = this.doc.documentElement;
+
+    htmlEl.classList.forEach((className) => {
+      if (className.endsWith(CLASS_SUFFIX)) {
+        this.renderer.removeClass(htmlEl, className);
+      }
     });
```

Comment:
Consider: Would it be easier/faster to just remove the light/dark theme classes directly?

```suggestion
    htmlEl.classList.remove(LIGHT_THEME_CLASS);
    htmlEl.classList.remove(DARK_THEME_CLASS);
```

This way it's not an O(n) check over all the classes on the element, and there's no risk of removing other classes which happen to use the `theme-*` prefix.